### ISSUE TITLE
Support defining "logical lines" for navigation purposes

### ISSUE THREAD

**Issue by Screwtapello** [mawww/kakoune#1917]
# Summary

Currently, Kakoune is fundamentally based around lines terminated by a single `\n` character. However, this is not the only possible structure for text-based data, and Kakoune would be more useful if it could be taught to redefine "lines".

# Motivation

Kakoune has a lot of features that work specifically with `\n`-separated lines, like the `12g` that moves the cursor to just after the 12th `\n` in the file, `j` and `k` that move the cursor to the same horizontal offset after or before the next or previous `\n` respectively, `o` that moves the cursor just past the next `\n`, inserts another `\n` and enters insert mode, that kind of thing.

However, while most text-files use `\n` to separate distinct data items, some text-files use larger or smaller-scale patterns, and some text-files use different characters entirely.

## Different separators

ASCII defines a suite of dedicated separator characters at the end of the C0 control-character block:

  - `\x1C` File Separator
  - `\x1D` Group Separator
  - `\x1E` Record Separator
  - `\x1F` Unit Separator

One of the problems with CSV files is that the characters used to separate fields and records (traditionally "comma" and "newline") are also quite likely to appear inside fields, requiring intricate quoting systems (plural, we wouldn't want to be *interoperable* now would we?).

Using "Unit Separator" and "Record Separator" instead would be ideal, except that these rarer characters aren't well-supported in text editors. If Kakoune allowed a buffer to use a custom "line" definition, this would be much simpler.

## Support editing logical lines

Many programming languages (including Kakoune's scripting language) support line-continuation, where a backslash (`\`) at the end of a line means the next line is a logical continuation rather than a new statement. For example, maybe the author wants to wrap a long line to fit a character limit, or to emphasise repeated text by lining it up vertically.

When one logical line covers multiple physical lines, this makes editing more complex. For example, let's say I have a bunch of mappings defined in my `kakrc`:

    map global normal \
        a %{:echo do a thing<ret>}
    map global normal \
        b %{:echo do b thing<ret>}
    map global normal \
        c %{:echo do c thing<ret>}
        
Let's say I decide "global" is too broad a scope, and these mappings should be window-specific. Today, the easiest approach is probably:

  - move to the first line
  - hit `X` five times to select all the lines containing scopes
  - Just select the scopes we care about with `sglobal<ret>`
  - Make the change with `cwindow<esc>`

If there were some way to teach Kakoune about logical lines, this could be as easy as:

  - move to the first line
  - hit `ww` to select `global`
  - hit `CCC` to select the same text on the next three lines
  - Make the change with `cwindow <esc>`

That's 14 keys instead of 21.

## Supporting multi-line records

Recently I've been working on a nice way to handle Rust error messages in Kakoune. I would have built on Kakoune's standard `make` integration, but unfortunately that code is heavily built on the assumption that each error will be printed on a single line. Rust's error messages look like this:

    error: Some error message goes here
      --> path/to/file:line:col
    <annotated source code with arrows pointing at things>
    
(Rust *does* support producing machine-readable errors, but by "machine readable" they mean "JSON", which doesn't help Kakoune much)

Following the model of `make.kak`, I made a `:cargo-jump` command to jump to the error under the cursor. However, if the cursor is on, say, the `-->` in the above example, then searching backward won't find that error (since it doesn't end before the cursor position) and searching forward won't find it either (since it doesn't begin after the cursor). As a workaround, to find the current error I search for the next error and then immediately search for the previous error.

On the other hand, if I were able to define a "line" as anything matching the error regex, then selecting the current error would be as easy as `x`, and next-error and previous-error would just be `j` and `k`.

## Supporting multi-record lines

Kakoune has excellent support for visually wrapping text to fit the width of the current window, but a common complaint is that keys like `j` and `k` only move according to physical lines, making it awkward to move the cursor into the middle of a paragraph.

If I were able to define a "line" with a regex like `[^\n]{1,79}` then I could approximate the navigational ease of hard-wrapped text while still actually editing soft-wrapped text.

# Design

Probably the most intuitive and flexible way to define consecutive, navigable regions of a file is with multiple selections. I'd love for the feature to look and work something like this:

  - I (or a plugin with a hook for the filetype option) selects all the logical lines in a file
  - A command "bakes" those selections, making them like a mask for future editing operations. Text outside the selections is dimmed and uneditable
  - In case these logical lines contain sub-structures (perhaps a record is big enough that each field should be presented as a line), more selections can be made and "baked", further restricting the available text.
  - It should also be possible to "unbake" logical lines, returning them to being a selection.

Basically, it's like a less-transient version of multiple selections, allowing a middle ground between "make every change individually" and "make all changes at once but every change site must be edited identically".

# Unresolved questions

  - What happens to text that's outside a designated "line"? Can we navigate to it, or is it immutable?
  - If we're using a custom line definition, what do commands like `o` and `<a-j>` do?
  - Would it be better to define logical lines by writing a regex to match line *separators* rather than lines?

**Comment by Delapouite** [comment]
In the _Design_ section of your comment the notion of "baking" selections and edit them one at a time reminds me a lot of occivink's phantom selections: https://github.com/occivink/kakoune-phantom-selection

Regarding the whole concept, I found it close to another idea I had in the back of my mind. Here's the hypothetical (and very rough) usage:
when you start Kakoune, you're in `char` mode. So `hjkl` keys use this *unit* to move and make the selections.
If you press `w` you're now in `word` mode. This time `hjkl` keys jumps from word to word. Press `<esc>` to go back to `char` mode and now press `p` (imagine it's not bound to paste). You're now in `paragraph` mode…etc

**Comment by Screwtapello** [comment]
Since I created this ticket I've been thinking more about it, and about how I use Kakoune versus other editors.

vi requires me to do bulk editing as a batch: I write a complex search-and-replace regex and apply it to the whole document, or I record a complex macro and repeatedly trigger it, and if I hit a problem I have to undo and try again from scratch.

Meanwhile, in Kakoune, I can write a simple regex with `%s`, and if it selects too much or too little, as long as the result has the right number of cursors I can patch it up manually before doing my next bulk action. Even if it has the wrong number of cursors, I might be able to patch it up with `<a-space>` to remove extra cursors or ctrl-click to add new ones. Despite Kakoune having "Modal editor" as the very first bullet-point on the homepage, this modelessness, this ability to mix automated changes and manual fix-up without losing state, is one of the best things about it.

But of course Kakoune isn't entirely modeless; if you're part-way through a change and you decide you need to adjust the document *content* rather than just the *selection*, you're out of luck. Perhaps I just need to get more familiar with `<z>`/`<Z>`... but since they overwrite a register instead of pushing to/popping from a stack, I'm sure that'll lead to heartache eventually.

**Comment by mawww** [comment]
I'll close this because I think it departs too much from Kakoune's design. Making line ending configurable means commands cannot be as predictible anymore, they depend on the current line ending setting, meaning we cannot do robust scripting with Kakoune's editing language without taking into account the various possible configurations.
